### PR TITLE
[web] Encode AssetManifest.bin as JSON and use that on the web.

### LINKED ISSUE #124883: Asset images are not loading (with Internet Download Manager installed on Microsoft Windows)
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Create a simple flutter project (Which supports web)
2. Just create an Image widget and load a png image from your assets
3. Run your application on chrome

### Expected results

I expected flutter to render my asset image normally.

### Actual results

Since I have been upgraded to the latest flutter version which is 3.10.0-6.0.pre.9 on master channel I am confronting this issue. First of all my asset images are not loading anywhere in the application.
This is a little bit odd but it might be helpful. When application runs and renders an Image widget which is loading from assets, my IDM application download dialog pops up and asks to download AssetManifest.bin.

**Note**: All assets images were loading in flutter v3.7.x and my images are not corrupted.

Here is the picture 
image

**Note**: I have IDM installed on my system all the time and this problem is happening after upgrading flutter to the latest version on master channel.

### Code sample

<details><summary>Code sample</summary>

```yaml
flutter:
  assets:
    - assets/images/logo.png
```

```dart
Image.asset('assets/images/logo.png', width: 85, height: 85, )
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

image

</details>

### Logs

<details><summary>Logs</summary>

```console
The following FormatException was thrown while resolving an image:
Message corrupted

When the exception was thrown, this was the stack: 
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 288:49  throw_
packages/flutter/src/services/message_codecs.dart 475:7                       readValue
packages/flutter/src/services/message_codecs.dart 342:28                      decodeMessage
packages/flutter/src/services/asset_manifest.dart 57:55                       fromStandardMessageCodecMessage
packages/flutter/src/services/asset_manifest.dart 59:3                        _#fromStandardMessageCodecMessage#tearOff
dart-sdk/lib/async/zone.dart 1661:54                                          runUnary
dart-sdk/lib/async/future_impl.dart 147:18                                    handleValue
dart-sdk/lib/async/future_impl.dart 784:44                                    handleValueCallback
dart-sdk/lib/async/future_impl.dart 813:13                                    _propagateToListeners
dart-sdk/lib/async/future_impl.dart 584:5                                     [_completeWithValue]
dart-sdk/lib/async/future_impl.dart 657:7                                     callback
dart-sdk/lib/async/schedule_microtask.dart 40:11                              _microtaskLoop
dart-sdk/lib/async/schedule_microtask.dart 49:5                               _startMicrotaskLoop
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 177:15           <fn>
Image provider: ScrollAwareImageProvider<Object>()
Image configuration: ImageConfiguration(bundle: PlatformAssetBundle#97f8e(), devicePixelRatio: 1.0, locale: en_US, textDirection: TextDirection.ltr, size: Size(135.0, 135.0), platform: windows)
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[√] Flutter (Channel master, 3.10.0-6.0.pre.9, on Microsoft Windows [Version 10.0.19045.2846], locale en-US)
    • Flutter version 3.10.0-6.0.pre.9 on channel master at C:\FlutterSdk\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 119744a63c (2 hours ago), 2023-04-14 10:43:17 -0700
    • Engine revision ea0e1b1700
    • Dart version 3.1.0 (build 3.1.0-4.0.dev)
    • DevTools version 2.23.1
    • Pub download mirror https://pub.dev

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at C:\AndroidSdk
    • Platform android-33, build-tools 33.0.1
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2022 17.3.5)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.3.32922.545
    • Windows 10 SDK version 10.0.20348.0

[√] Android Studio (version 2021.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[√] VS Code (version 1.77.0)
    • VS Code at C:\Users\MonoBit\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.62.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19045.2846]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 112.0.5615.49
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 112.0.1722.39

[√] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by ditman**
This PR modifies the web build slightly to create an `AssetManifest.json`, that is a JSON(base64)-encoded version of the `AssetManifest.bin` file.

_(This should enable all browsers to download the file without any interference, and all servers to serve it with the correct headers.)_

It also modifies Flutter's `AssetManifest` class so it loads and uses said file `if (kIsWeb)`.

### Issues

* Fixes https://github.com/flutter/flutter/issues/124883

### Tests

* Unit tests added.
* Some tests that run on the Web needed to be informed of the new filename, but their behavior didn't have to change (binary contents are the same across all platforms).
* I've deployed a test app, so users affected by the BIN issue may take a look at the PR in action:
  * https://dit-tests.web.app

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by jonahwilliams** [comment]
base64 decoding is pretty slow. Have you benchmarked this versus the new binary for and/or the old JSON format?

**Comment by ditman** [comment]
> Have you benchmarked this versus the new binary for and/or the old JSON format?

@jonahwilliams Nope, do those benchmarks exist or do they need to be written?

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 1]
```diff
@@ -21,6 +23,14 @@ abstract class AssetManifest {
   /// Loads asset manifest data from an [AssetBundle] object and creates an
   /// [AssetManifest] object from that data.
   static Future<AssetManifest> loadFromAssetBundle(AssetBundle bundle) {
+    if (kIsWeb) {
+      return bundle.loadStructuredData(_kAssetManifestWebFilename, (String jsonData) async {
+        // Decode the manifest JSON file to the underlying BIN, and convert to ByteData.
+        final ByteData message = ByteData.sublistView(base64.decode(json.decode(jsonData) as String));
+        // Now we can keep operating as usual.
+        return _AssetManifestBin.fromStandardMessageCodecMessage(message);
```
Comment (yjbanov) [inline packages/flutter/lib/src/services/asset_manifest.dart:31]: As a quick sanity check, can you use `FlutterTimeline` or something to time how long it takes to decode `jsonData` into an instance of `_AssetManifestBin`?

Comment (ditman) [reply packages/flutter/lib/src/services/asset_manifest.dart:31]: Definitely. I've also redeployed the app in the description in `--profile` mode so we can look at the flame chart from chrome.

**Comment by jonahwilliams** [comment]
See https://github.com/flutter/flutter/blob/master/dev/benchmarks/microbenchmarks/lib/foundation/decode_and_parse_asset_manifest.dart

**Comment by jonahwilliams** [comment]
This change will also still require us to make an additional network requst before any assets can be downloaded.

**Comment by ditman** [comment]
@jonahwilliams @yjbanov

I took the microbenchmark that Jonah suggested [here](https://github.com/flutter/flutter/pull/131382#issuecomment-1652876205) (thanks!), and modified it slightly to run on the web.

I had to:

* Add a `web/index.html` (I actually did `flutter create --platform web .`)
* Changed the number of iterations from 1000 to 10000 (with 1k, each time I refreshed the benchmark, `master` would be even faster, so I assumed this was the browser warming up. I raised the number of iterations to try to give a chance to all benchmarks to be "similarly warm" without having to run them several times).

I ran the benchmark in `master` and the 2 PRs. I ran each example like this:

```
refresh-tools && flutter clean && flutter run -d web-server --target lib/foundation/decode_and_parse_asset_manifest.dart --web-port 7357 --release
```

(`refresh-tools` is a small command that I have to delete the `flutter_tools` cache, to ensure that changes actually work)

Here's the results after 10k iterations:

| master | #130419 | #131382 (this) |
|--------|---------|----------------|
| master |  |  |
| 57078ms | 58278ms | 59956ms |

I only ran each microbenchmark once in my gLinux machine, so I take the numbers with a healthy amount of skepticism. Squinting a little bit, we could say that "this" is 2878ms slower than `master`, so I guess downloading a ~30% larger file and json-base64-decoding it incurs in an (average(, not the best for benchmarks)) ~0.3ms penalty.

It also seems that #130419 is still downloading the `AssetManifest.bin` on the web, which is hurting it badly in this benchmark, I expected it to wipe the floor with the other two branches (I don't know exactly if I messed up here, but it seems that the global window array is being set, so changes are being applied. I can also take a look there and see what's up.)

**Comment by jonahwilliams** [comment]
FYI @andrewkolos maybe you can help out with the benchmark issue here? Do we need to use a different method to load and parse the manifest?

**Comment by yjbanov** [comment]
Thanks for benchmarking this, @ditman! Looks like there's no material performance difference between these approaches. If so, I'd say let's go with the simplest solution.

**Comment by ditman** [comment]
@yjbanov Andrew's PR should be much, much faster, I'm shaving that yak now to find why is it downloading the manifest (when it's set in the DOM already)

**Comment by ditman** [comment]
I pushed a [fix](https://github.com/flutter/flutter/pull/130419/commits/dd6ce29b8387cced79df0508284811857bcc489c) to @andrewkolos' PR, and NOW it's chugging!

Screenshot 2023-07-27 at 3 46 13 PM

 11ms for 10k iterations of parsing the manifest (now that the bytes are getting inlined, there's not much work to do for the loader, other than juggling the bytes coming from the JS global).

/cc @jonahwilliams @yjbanov.

**Comment by jonahwilliams** [comment]
What are the numbers here? Sorry its a bit hard to follow the different issues/threads.

**Comment by andrewkolos** [comment]
Heya, sorry for the silence--I was out at the tail end of last week.

I initially considered generating a separate version of the asset manifest for web, but I wanted to cut out the network request for the asset manifest to speed up (initial) asset loading and avoid having all pending asset loads depending on a single network request). However, if we are willing to delay addressing these properly (i.e. fixing #118115), I don't mind this adopting this solution. The other reason I avoided a solution like this one is that I wanted to avoid having to write/maintain a bunch of logic to construct/consume a JSON file, but your solution of just taking the existing bin data and encoding it as a string is pretty clever and doesn't strike me as _too_ slow or data-consuming.

Barring strong opinions from others, I'd like to adopt this PR and close my original PR (after a few days or so).

**Comment by ditman** [comment]
Thanks @andrewkolos! I'll take a second look at the tests I broke here and get this one ready to land. I'd like to give this version a shot before we do drastic changes to the build process, but when we do those, we'll be sure to include both the font manifests and the assets one!

**Comment by ditman** [comment]
Agh, I was forgetting to pass `--platform=chrome` to my `flutter test` commands, that's why things were working locally but still failing in CI!

**Comment by ditman** [comment]
@andrewkolos offered to prepare a g3Fix for this so the "Google Testing" check passes and the PR can land. (It may be *this* PR that one that lands, or some other that @andrewkolos owns so he can do any changes he needs!)

**Comment by andrewkolos** [comment]
One challenge I failed to foresee with this is the amount of change that might be required in g3. Unlike past asset manifest changes, this adds a file that is _conditionally_ included in the build. This makes updating tests and build rules more difficult. In addition (though not as important), these kind of changes tend to break tests for customers that mock/fake `AssetBundle` in their tests. One of the reasons I initially liked this PR over my initial PR was it's relative simplicity. However, when considering g3 and testing, it may not be so simple.

It's possible that updating g3 to work with this PR might not be _that_ complicated, but there are unknowns. Given that I've already invested at least 3x as much time as I wanted to on this issue, I'd like to go with the option that will probably have the fewest surprises. As a result, I may revisit my previous PR (https://github.com/flutter/flutter/pull/130419).

**Comment by ditman** [comment]
> It's possible that updating g3 to work with this PR might not be that complicated, but there are unknowns.

OK, I guess I'll give that a shot too.

**Comment by andrewkolos** [comment]
> > It's possible that updating g3 to work with this PR might not be that complicated, but there are unknowns.
> 
> ~OK, I guess I'll give that a shot too.~
> 
> @andrewkolos is right, this is a world of hurt internally, what a shame :/

Do you have opinions against the core idea of https://github.com/flutter/flutter/pull/130419? As far as I can tell, @jonahwilliams and @yjbanov are satisfied with the core idea, as long as it uses a public dart API (as opposed to a hidden JS global). I really don't want to foist a bunch of g3 work on you. I'm happy to rework #130419 instead (though I can only hope g3 doesn't use custom code for generating the web entry point 🙃--I never got to find out because FroB bugged out and refused to test the PR).

**Comment by ditman** [comment]
I have a g3fix ready for this, applying.

**Comment by ditman** [comment]
> I can only hope g3 doesn't use custom code for generating the web entry point 🙃

@andrewkolos sorry, I've just seen this message 😅

In fact, g3 does uses custom code to generate the web entrypoint (see `cs//depot/google3/mobile/flutter/build_defs/internal/web_rules.bzl;l=195`), so you'd still have to go through the g3fix process to land the other version of the change.

I prefer this PR because it keeps about the same complexity that we currently have in `master`, fixes the problem that some people have, and gives us some extra time to actually design what's going to be our web inlining strategy. (As Jonah said, it's not the responsibility of your PR to find that solution.)

I guess I'm trying to avoid having one-offs for each file we may want to inline, and find a consolidated solution. What if tomorrow we want to make the FontManifest.json a binary file? (I'd rather it became a .bin.json for initial web solution, than creating another different setter to the engine, etc...)

**Comment by ditman** [comment]
Google testing should pass with this g3fix: cl/553547021

**Comment by christopherfujino** [comment]
@ditman looks like the testing cl didn't apply your g3fix: cl/553348554. Should we retry?

**Comment by christopherfujino** [comment]
friendly ping @ditman

**Comment by ditman** [comment]
(Apologies, I've been OOO for the last few weeks!)

@christopherfujino Yes please!

This PR has supposedly passed a TGP run (cl/553273631), but I never got it to run alongside its g3fix (cl/553547021).

If there's anything you can do (or that I need to do) to move things along, I'd love to get it done!

**Comment by christopherfujino** [comment]
> (Apologies, I've been OOO for the last few weeks!)
> 
> @christopherfujino Yes please!
> 
> This PR has supposedly passed a TGP run (cl/553273631), but I never got it to run alongside its g3fix (cl/553547021).
> 
> If there's anything you can do (or that I need to do) to move things along, I'd love to get it done!

Well unrelated, Google testing is borked: https://github.com/flutter/flutter/issues/133814

**Comment by ditman** [comment]
😿

**Comment by ditman** [comment]
Looks like the FROB test passed, yay! Can you PTAL @yjbanov?

File: packages/flutter/lib/src/services/asset_manifest.dart [hunk 2]
```diff
@@ -2,25 +2,37 @@
...
+const String _kAssetManifestWebFilename = 'AssetManifest.bin.json';
+
 /// Contains details about available assets and their variants.
 /// See [Resolution-aware image assets](https://docs.flutter.dev/ui/assets-and-images#resolution-aware)
 /// to learn about asset variants and how to declare them.
 abstract class AssetManifest {
   /// Loads asset manifest data from an [AssetBundle] object and creates an
   /// [AssetManifest] object from that data.
   static Future<AssetManifest> loadFromAssetBundle(AssetBundle bundle) {
+    if (kIsWeb) {
```
Comment (yjbanov) [inline packages/flutter/lib/src/services/asset_manifest.dart:28]: nit: a couple of lines (or links) above the `if` explaining how/why things are different on the web would be useful to future contributors.

**Review by yjbanov — APPROVED** [review]

**Comment by ditman** [comment]
Working on addressing yegor's nits (also giving some internal code time to land before landing this fix!)

File: packages/flutter/lib/src/services/asset_manifest.dart (hunk 2 above)
Comment (ditman) [reply packages/flutter/lib/src/services/asset_manifest.dart:33]: (Done)

**Comment by ditman** [comment]
Adding `autosubmit`, 🤞

**Comment by fhuonder** [comment]
Hi @ditman 

Do you know when this fix will land on a stable release?
We currently have the problem that in some companies the firewalls block the bin-file of our flutter-web-app.

Regards,
Florian
